File: server/src/routes/projects.js
const express = require('express');
const { authMiddleware } = require('../middleware/auth');
const { validateProject } = require('../middleware/validators');
const { runAsync, getAsync, allAsync } = require('../db');
const { invalidateUserCache, invalidateCache } = require('../cache');
const logger = require('../logger');

const router = express.Router();

router.use(authMiddleware);

router.get('/', async (req, res) => {
  try {
    const projects = await allAsync(
      'SELECT id, name, description, created_at, updated_at FROM projects WHERE user_id = ? ORDER BY updated_at DESC',
      [req.userId]
    );

    res.json({
      success: true,
      data: projects
    });
  } catch (error) {
    console.error('Get projects error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch projects'
    });
  }
});

router.post('/', validateProject, async (req, res) => {
  try {
    const { name, description } = req.body;

    const result = await runAsync(
      'INSERT INTO projects (user_id, name, description) VALUES (?, ?, ?)',
      [req.userId, name, description || '']
    );

    invalidateUserCache(req.userId);
    logger.info(`Project created: ${name} (user: ${req.userId})`);

    res.status(201).json({
      success: true,
      message: 'Project created successfully',
      data: {
        id: result.id,
        name,
        description: description || '',
        created_at: new Date().toISOString()
      }
    });
  } catch (error) {
    logger.error('Create project error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to create project'
    });
  }
});

router.get('/:projectId', async (req, res) => {
  try {
    const { projectId } = req.params;

    const project = await getAsync(
      'SELECT id, name, description, created_at, updated_at FROM projects WHERE id = ? AND user_id = ?',
      [projectId, req.userId]
    );

    if (!project) {
      return res.status(404).json({
        success: false,
        error: 'Project not found'
      });
    }

    res.json({
      success: true,
      data: project
    });
  } catch (error) {
    console.error('Get project error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch project'
    });
  }
});

router.put('/:projectId', validateProject, async (req, res) => {
  try {
    const { projectId } = req.params;
    const { name, description } = req.body;

    const project = await getAsync(
      'SELECT id FROM projects WHERE id = ? AND user_id = ?',
      [projectId, req.userId]
    );

    if (!project) {
      return res.status(404).json({
        success: false,
        error: 'Project not found'
      });
    }

    await runAsync(
      'UPDATE projects SET name = ?, description = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?',
      [name, description || '', projectId]
    );

    res.json({
      success: true,
      message: 'Project updated successfully',
      data: {
        id: projectId,
        name,
        description: description || '',
        updated_at: new Date().toISOString()
      }
    });
  } catch (error) {
    console.error('Update project error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update project'
    });
  }
});

router.delete('/:projectId', async (req, res) => {
  try {
    const { projectId } = req.params;

    const project = await getAsync(
      'SELECT id FROM projects WHERE id = ? AND user_id = ?',
      [projectId, req.userId]
    );

    if (!project) {
      return res.status(404).json({
        success: false,
        error: 'Project not found'
      });
    }

    await runAsync(
      'DELETE FROM projects WHERE id = ?',
      [projectId]
    );

    res.json({
      success: true,
      message: 'Project deleted successfully'
    });
  } catch (error) {
    console.error('Delete project error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete project'
    });
  }
});

module.exports = router;

File: Applications/ServerGo/internal/handlers/projects.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"github.com/SandwichEater577/valyxo-backend/internal/db"
	"github.com/SandwichEater577/valyxo-backend/internal/middleware"
	"github.com/gin-gonic/gin"
)

// ProjectRequest represents project input
type ProjectRequest struct {
	Name        string `json:"name" binding:"required,min=1,max=100"`
	Description string `json:"description"`
}

// GetProjects returns all projects for current user
func GetProjects(c *gin.Context) {
	userID := middleware.GetUserID(c)

	projects, err := db.GetProjectsByUserID(userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to fetch projects",
		})
		return
	}

	if projects == nil {
		projects = []db.Project{}
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    projects,
	})
}

// GetProject returns a single project
func GetProject(c *gin.Context) {
	userID := middleware.GetUserID(c)
	projectID, err := strconv.ParseInt(c.Param("projectId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Invalid project ID",
		})
		return
	}

	project, err := db.GetProjectByID(projectID, userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to fetch project",
		})
		return
	}
	if project == nil {
		c.JSON(http.StatusNotFound, gin.H{
			"success": false,
			"error":   "Project not found",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"data":    project,
	})
}

// CreateProject creates a new project
func CreateProject(c *gin.Context) {
	userID := middleware.GetUserID(c)

	var req ProjectRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Validation error: " + err.Error(),
		})
		return
	}

	projectID, err := db.CreateProject(userID, req.Name, req.Description)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to create project",
		})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"success": true,
		"message": "Project created successfully",
		"data": gin.H{
			"id":          projectID,
			"name":        req.Name,
			"description": req.Description,
			"created_at":  time.Now().Format(time.RFC3339),
		},
	})
}

// UpdateProject updates an existing project
func UpdateProject(c *gin.Context) {
	userID := middleware.GetUserID(c)
	projectID, err := strconv.ParseInt(c.Param("projectId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Invalid project ID",
		})
		return
	}

	// Check ownership
	project, err := db.GetProjectByID(projectID, userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to fetch project",
		})
		return
	}
	if project == nil {
		c.JSON(http.StatusNotFound, gin.H{
			"success": false,
			"error":   "Project not found",
		})
		return
	}

	var req ProjectRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Validation error: " + err.Error(),
		})
		return
	}

	if err := db.UpdateProject(projectID, req.Name, req.Description); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to update project",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "Project updated successfully",
		"data": gin.H{
			"id":          projectID,
			"name":        req.Name,
			"description": req.Description,
			"updated_at":  time.Now().Format(time.RFC3339),
		},
	})
}

// DeleteProject deletes a project
func DeleteProject(c *gin.Context) {
	userID := middleware.GetUserID(c)
	projectID, err := strconv.ParseInt(c.Param("projectId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Invalid project ID",
		})
		return
	}

	// Check ownership
	project, err := db.GetProjectByID(projectID, userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to fetch project",
		})
		return
	}
	if project == nil {
		c.JSON(http.StatusNotFound, gin.H{
			"success": false,
			"error":   "Project not found",
		})
		return
	}

	if err := db.DeleteProject(projectID); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to delete project",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "Project deleted successfully",
	})
}
